server/src/main/kotlin/fi/oph/kitu/yki/YkiService.kt

```diff
@@ -16,20 +17,27 @@ class YkiService(
 ) {
     private val logger: Logger = LoggerFactory.getLogger(javaClass)
 
-    fun importYkiSuoritukset() {
-        val suoritus =
+    fun importYkiSuoritukset(
+        lastSeen: LocalDate? = null,
+        dryRun: Boolean? = null,
+    ): Int {
+        val suoritukset =
             ykiRestClient
                 .get()
-                .uri("/yki")
+                .uri("suoritukset")
                 .retrieve()
-                .body<YkiSuoritus>()
+                .body(String::class.java)
+                ?.asCsv<YkiSuoritus>() ?: throw RestClientException("Response body is empty")
 
-        if (suoritus == null) {
-            logger.info("No YKI suoritus found")
-            return
+        if (suoritukset.isEmpty()) {
+            throw RestClientException("The response is empty")
         }
 
-        repository.insertSuoritus(suoritus)
-        logger.info("YKI Suoritus added to repository")
+        if (dryRun == true) {
+            return 0
+        }
```

Comment:
Logituksen siirtäminen ylätasolle aiheuttaa nyt sellaisen ikävän ongelman, ettei täältä `importYkiSuoritukset` sisälmyksistä pysty käytännössä vaikuttamaan kuin yhteen avain-arvopariin (paluuarvo päätyy logitettavan `importSize` arvoksi).

Tavallaan haluaisin log eventteihin mieluummin enemmän kuin vähemmän tietoa. Esimerkiksi täältä olisi mukava logittaa `suoritukset.size` sellaisenaan, sen lisäksi että logitetaan lopussa montako tietuetta oikeasti tuotiin ja tallennettiin.

Haluaisin mieluummin nähdä yksittäisiä paljon dataa sisältäviä lokirivejä, kuin ylätasolla kirjattuja "onnistui" / "ei onnistunut" tyyppisiä.

Eli import-proseduurin aikana vain kerätään lokiriville tietoa avain-arvopareina ja lopussa logitetaan koko proseduurin lopputulos yhtenä rivinä. En tiedä onko käytetyissä kirjastoissa mitä tapoja tähän.

Yksi tapa voisi esim. olla, että metodiin passattaisiin `LoggingEventBuilder` sisään ja ylätasolla olisi käärivä try-finally, jonka finally-lohkossa logitettaisiin geneerinen `"import done"`. Kaikki muu tieto (onnistuiko operaatio, oliko dry-run, montako suoritusta saatiin API:sta, montako tietuetta tallennettiin, jne.) voitaisiin liittää rakenteellisesti avain-arvopareina metodille annettuun event builderiin.

Karkeasti siis:
```kotlin
// ScheduledTasks.kt
fun scheduledImportFoo() {
    // Luodaan log event
    var event = logger.atInfo()
    try {
        // Import-proseduuri saa eventin parametrina
        service.importFoo(event)
        
        // Importin jälkeen voidaan liittää vielä lisää dataa
        event
            .addKeyValue("success", true)
            .setMessage("import done successfully")
    }
    catch (e) {
        // Virhetilanteessa voidaan liittää toisenlaista dataa
        event
            .addKeyValue("success", false)
            .setMessage("import failed")
    }
    finally {
        // Myös muualla voidaan liittää lisäinfoa
        event
            .addKeyValue("foo", "bar")
            // Lopuksi logitetaan event yhtenä rakenteellisena möllykkänä
            .log()
    }
}

// Service.kt
fun importFoo(event) {
    // Eventtiin voidaan liittää lisäinfoa
    event.addKeyValue("isDryRun", isDryRun)

    val foos = fetchAllFoos()
    event.addKeyValue("fetchedFooCount", foos.size)
    
    if (isDryRun) {
        return
    }

    val importedFoos = repository.insertFoos(foos)
    event.addKeyValue("importedFooCount", importedFoos.size)
}
```

Eli ajatuksena kääntää logittaminen hieman päälaelleen; sen sijaan että logitetaan usein ja tarkkaa tilatietoa, kerätään koko proseduurin suoritus yhteen eventtiin, johon liitetään proseduurin edetessä lisäinfoa, eli logitetaan vain yksi rivi, joka sisältää rakenteellisesti kaiken olennaisen tiedon.

Näin lokirivin viestikenttä on enemmänkin vain vastaus kysymykseen "mitä yritettiin tehdä" ja kaikki yksityiskohdat siitä "mitä oikeasti tapahtui" löytyvät lokirivin rakenteellisesta datasta.

Reply:
Ylläoleva kommentti on ehkä hieman enemmän yleisen tason keskustelunavaus siitä, miten logitusta pitäisi tehdä, kuin vaatimus muutoksille tähän pullaan.